src/tests/utests/iotc_utest_memory_calloc.c

```diff
@@ -0,0 +1,195 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "iotc_macros.h"
+#include "iotc_tt_testcase_management.h"
+#include "tinytest.h"
+#include "tinytest_macros.h"
+
+#include "iotc.h"
+#include "iotc_allocator.h"
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+#ifndef IOTC_TT_TESTCASE_ENUMERATION__SECONDPREPROCESSORRUN
+
+#endif  // IOTC_TT_TESTCASE_ENUMERATION__SECONDPREPROCESSORRUN
+
+IOTC_TT_TESTGROUP_BEGIN(utest_memory_calloc)
+
+/* Happy cases. */
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__num_1__size_8,
+    {
+      void* ptr = __iotc_calloc( /*num=*/1, /*size=*/8);
+      tt_want_ptr_op(NULL, !=, ptr);
+      __iotc_free(ptr);
+    })
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc___num_255__size_55,
+    {
+      void* ptr = __iotc_calloc( /*num=*/255, /*size=*/55);
+      tt_want_ptr_op(NULL, !=, ptr);
+      __iotc_free(ptr);
+    })
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc___num_1000000__size_88,
+    {
+      void* ptr = __iotc_calloc( /*num=*/1000000, /*size=*/88);
+      tt_want_ptr_op(NULL, !=, ptr);
+      __iotc_free(ptr);
+    })
+
+/* Edge cases. */
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__num_0__size_8,
+    {
+      void* ptr = __iotc_calloc( /*num=*/0, /*size=*/8);
+      tt_want_ptr_op(NULL, ==, ptr);
+      __iotc_free(ptr);
+    })
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__num_1__size_0,
+    {
+      void* ptr = __iotc_calloc( /*num=*/1, /*size=*/0);
+      tt_want_ptr_op(NULL, ==, ptr);
+      __iotc_free(ptr);
+    })
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__num_0__size_0,
+    {
+      void* ptr = __iotc_calloc( /*num=*/0, /*size=*/0);
+      tt_want_ptr_op(NULL, ==, ptr);
+      __iotc_free(ptr);
+    })
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__num_max__size_0,
+    {
+      void* ptr = __iotc_calloc( /*num=*/SIZE_MAX, /*size=*/0);
+      tt_want_ptr_op(NULL, ==, ptr);
+      __iotc_free(ptr);
+    })
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__num_0__size_max,
+    {
+      void* ptr = __iotc_calloc( /*num=*/0, /*size=*/SIZE_MAX);
+      tt_want_ptr_op(NULL, ==, ptr);
+      __iotc_free(ptr);
+    })
+
+/* Edge cases - SIZE_MAX
+   Note: FreeRTOS heap size isn't large enough for SIZE_MAX allocations,
+   so these tests are disabled for those targets. */
+#ifndef IOTC_PLATFORM_IS_FREERTOS
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__num_1__size_max,
+    {
+      void* ptr = __iotc_calloc( /*num=*/1, /*size=*/SIZE_MAX);
+      tt_want_ptr_op(NULL, ==, ptr);
+      __iotc_free(ptr);
+    })
+
+IOTC_TT_TESTCASE(
+    utest__iotc_memory_calloc__num_max_minus_1__size_1,
+    {
+      void* ptr = __iotc_calloc( /*num=*/SIZE_MAX-1, /*size=*/1);
+      tt_want_ptr_op(NULL, ==, ptr);
```

Comment:
Hmm, still cannot allocate `SIZE_MAX - 1`?

Reply:
Nope.   FreeRTOS default heap size is much smaller:

In FreeRTOSConfig.h:
```
#define configTOTAL_HEAP_SIZE ((size_t)(350 * 1024))
```

Reply:
Oh I see what you're aying. I think this errro is less of "what does calloc due for performance" and more about how much heap the allocator can put aside in one contiguous. block.   Seems like it can't do SIZE_MAX-1 either, even on my mac and linux boxes.

Reply:
👌  thx, makes sense.

Reply:
SIZE_MAX is more than 18M TerraBytes... 🤦‍♂️

Reply:
so we can assume here it returns NULL pointer here... for years.